// react-kawaii/lib/es/components/backpack/Backpack.js
import React from 'react';
import PropTypes from 'prop-types';
import paths from './paths';
import getUniqueId from '../../utils/getUniqueId';
import Face from '../common/face/Face';
import Wrapper from '../common/wrapper/Wrapper';

var Backpack = function Backpack(_ref) {
  var size = _ref.size,
      color = _ref.color,
      mood = _ref.mood,
      className = _ref.className;
  return /*#__PURE__*/React.createElement(Wrapper, {
    className: className
  }, /*#__PURE__*/React.createElement("svg", {
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    width: size * 0.63,
    height: size,
    viewBox: "0 0 138 218"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("path", {
    id: "pocket__shape-path",
    d: paths.pocketShapePath
  }), /*#__PURE__*/React.createElement("path", {
    id: "pocket__shape__shadow-path",
    d: paths.pocketShapeShadowPath
  })), /*#__PURE__*/React.createElement("g", {
    id: "kawaii-backpack",
    transform: "translate(0 -1)"
  }, /*#__PURE__*/React.createElement("g", {
    id: "right-shoulder-strap",
    transform: "matrix(-1 0 0 1 125.901 13.054)"
  }, /*#__PURE__*/React.createElement("path", {
    id: "right-shoulder-strap__back",
    fill: color,
    d: "M0 5.948v43.619l13.879-14.792z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "right-shoulder-strap__back__shadow",
    fill: "#FFF",
    fillRule: "nonzero",
    opacity: "0.4",
    d: "M0 5.948v43.619l13.879-14.792z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "right-shoulder-strap__front",
    fill: color,
    d: "M9.193 34.697L0 5.127 21.014 0l7.735 24.677z"
  })), /*#__PURE__*/React.createElement("g", {
    id: "left-shoulder-strap",
    transform: "translate(11.896 13.054)"
  }, /*#__PURE__*/React.createElement("path", {
    id: "left-shoulder-strap__back",
    fill: color,
    d: "M0 5.948v43.619l13.879-14.792z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "left-shoulder-strap__back_shadow",
    fill: "#FFF",
    fillRule: "nonzero",
    opacity: "0.4",
    d: "M0 5.948v43.619l13.879-14.792z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "left-shoulder-strap__front",
    fill: color,
    d: "M9.193 34.697L0 5.127 21.014 0l7.735 24.677z"
  })), /*#__PURE__*/React.createElement("g", {
    id: "hanging-loop",
    transform: "translate(47.585)"
  }, /*#__PURE__*/React.createElement("path", {
    id: "hanging-loop__back",
    fill: color,
    d: "M16.853 7.931h9.913L21.81.991z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "hanging-loop__back_shadow",
    fill: "#121212",
    fillRule: "nonzero",
    opacity: "0.14",
    d: "M16.853 7.931h9.913L21.81.991z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "hanging-loop__shape",
    fill: color,
    d: paths.hangingLoopShape
  })), /*#__PURE__*/React.createElement("g", {
    id: "main-compartment",
    transform: "translate(0 26.617)"
  }, /*#__PURE__*/React.createElement("path", {
    id: "main-compartment__shape",
    fill: color,
    d: paths.mainCompartmentShape
  }), /*#__PURE__*/React.createElement("path", {
    id: "main-compartment__bottom-stripe",
    fill: "#121212",
    fillRule: "nonzero",
    d: paths.mainCompartmentBottomStripe,
    opacity: "0.07"
  }), /*#__PURE__*/React.createElement(Face, {
    mood: mood,
    transform: "translate(36 70)",
    uniqueId: getUniqueId()
  }), /*#__PURE__*/React.createElement("path", {
    id: "main-compartment__top",
    fill: "#121212",
    fillRule: "nonzero",
    d: paths.mainCompartmentTop,
    opacity: "0.07"
  })), /*#__PURE__*/React.createElement("g", {
    id: "pocket",
    transform: "translate(12.888 135.981)"
  }, /*#__PURE__*/React.createElement("mask", {
    id: "mask-4",
    fill: "#fff"
  }, /*#__PURE__*/React.createElement("use", {
    xlinkHref: "#pocket__shape-path"
  })), /*#__PURE__*/React.createElement("use", {
    id: "pocket__shape",
    fill: color,
    xlinkHref: "#pocket__shape-path"
  }), /*#__PURE__*/React.createElement("mask", {
    id: "mask-6",
    fill: "#fff"
  }, /*#__PURE__*/React.createElement("use", {
    xlinkHref: "#pocket__shape__shadow-path"
  })), /*#__PURE__*/React.createElement("use", {
    id: "pocket__shape__shadow",
    fill: "#121212",
    fillRule: "nonzero",
    opacity: "0.14",
    xlinkHref: "#pocket__shape__shadow-path"
  }), /*#__PURE__*/React.createElement("g", {
    id: "Group-6",
    mask: "url(#mask-6)"
  }, /*#__PURE__*/React.createElement("g", {
    transform: "translate(0 15.862)"
  }, /*#__PURE__*/React.createElement("path", {
    id: "Rectangle-7",
    fill: "#33363B",
    opacity: "0.408",
    d: "M0 2.974h113.014v2.974H0z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "Rectangle-7-Copy",
    fill: "#555D67",
    opacity: "0.408",
    d: "M0 0h113.014v2.974H0z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "Rectangle-8",
    fill: "#6D727A",
    opacity: "0.587",
    d: "M107.066 0h5.948v5.948h-5.948z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "zip",
    fill: "#49505A",
    d: paths.zip,
    opacity: "0.746"
  }))), /*#__PURE__*/React.createElement("g", {
    id: "brand",
    mask: "url(#mask-6)"
  }, /*#__PURE__*/React.createElement("g", {
    transform: "translate(78.737 39.368)"
  }, /*#__PURE__*/React.createElement("path", {
    id: "Rectangle-6",
    fill: "#FFFDFD",
    d: "M0 0h27.789v18.526H0z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "Rectangle-6-Copy",
    fill: "#121212",
    opacity: "0.298",
    d: "M1.158 1.158h25.474v16.211H1.158z"
  }), /*#__PURE__*/React.createElement("path", {
    id: "brand",
    fill: "#FFF",
    d: "M2.316 2.316h23.158v13.895H2.316z"
  })))))));
};

Backpack.propTypes = {
  /**
   * Size of the width
   * */
  size: PropTypes.number,
  mood: PropTypes.oneOf(['sad', 'shocked', 'happy', 'blissful', 'lovestruck', 'excited', 'ko']),

  /**
   * Hex color
   */
  color: PropTypes.string
};
Backpack.defaultProps = {
  size: 240,
  mood: 'blissful',
  color: '#FFD882'
};
export default Backpack;
