// react-kawaii/lib/native/components/creditCard/CreditCard.js
import React from 'react';
import PropTypes from 'prop-types';
import Svg, { G, Path } from 'react-native-svg';
import paths from './paths';
import Face from '../common/face/Face';
import getUniqueId from '../../utils/getUniqueId';
import Wrapper from '../common/wrapper/Wrapper';

var CreditCard = function CreditCard(_ref) {
  var size = _ref.size,
      color = _ref.color,
      mood = _ref.mood,
      className = _ref.className;
  return /*#__PURE__*/React.createElement(Wrapper, {
    className: className
  }, /*#__PURE__*/React.createElement(Svg, {
    width: size * 1.38,
    height: size,
    viewBox: "0 0 198 143",
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink"
  }, /*#__PURE__*/React.createElement(G, {
    id: "kawaii-creditCard"
  }, /*#__PURE__*/React.createElement(G, {
    id: "kawaii-creditCard__body",
    fillRule: "nonzero"
  }, /*#__PURE__*/React.createElement(Path, {
    d: paths.shape,
    id: "kawaii-creditCard__shape",
    fill: color
  }), /*#__PURE__*/React.createElement(Path, {
    d: paths.shadow,
    id: "kawaii-creditCard__shadow",
    fill: "#000",
    opacity: ".1"
  }), /*#__PURE__*/React.createElement(Path, {
    id: "kawaii-creditCard__stripe",
    fill: "#000",
    d: "M0 17h198v27H0z"
  })), /*#__PURE__*/React.createElement(Face, {
    mood: mood,
    transform: "translate(66 73)",
    uniqueId: getUniqueId()
  }))));
};

CreditCard.propTypes = {
  size: PropTypes.number,
  mood: PropTypes.oneOf(['sad', 'shocked', 'happy', 'blissful', 'lovestruck', 'excited', 'ko']),
  color: PropTypes.string
};
CreditCard.defaultProps = {
  size: 200,
  mood: 'blissful',
  color: '#83D1FB'
};
export default CreditCard;
